ykrt/src/compile/jitc_yk/trace_builder.rs

```diff
@@ -94,51 +91,49 @@ impl TraceBuilder {
+        // Find the control point call to retrieve the live variables from its safepoint.
         //
         // FIXME: Stash the location at IR lowering time, instead of searching at runtime.
+        let mut safepoint = None;
         let mut inst_iter = blk.insts.iter().enumerate().rev();
         for (_, inst) in inst_iter.by_ref() {
             // Is it a call to the control point, then insert loads for the trace inputs. These
             // directly reference registers or stack slots in the parent frame and thus don't
             // necessarily result in machine code during codegen.
             if inst.is_control_point(self.aot_mod) {
-                let safepoint = inst.safepoint().unwrap();
-                let (rec, _) = AOT_STACKMAPS
-                    .as_ref()
-                    .unwrap()
-                    .get(usize::try_from(safepoint.id).unwrap());
-
-                debug_assert!(safepoint.lives.len() == rec.live_vars.len());
-                for idx in 0..safepoint.lives.len() {
-                    let aot_op = &safepoint.lives[idx];
-                    let input_tyidx = self.handle_type(aot_op.type_(self.aot_mod))?;
-                    let load_ti_inst =
-                        jit_ir::LoadTraceInputInst::new(u32::try_from(idx).unwrap(), input_tyidx)
-                            .into();
-                    self.jit_mod.push(load_ti_inst)?;
-
-                    // Get the location for this input variable.
-                    let var = &rec.live_vars[idx];
-                    if var.len() > 1 {
-                        todo!();
-                    }
-                    self.jit_mod.push_tiloc(var.get(0).unwrap().clone());
-                    self.local_map.insert(
-                        aot_op.to_inst_id(),
-                        jit_ir::Operand::Var(self.jit_mod.last_inst_idx()),
-                    );
-                    self.jit_mod
-                        .push_loop_start_var(jit_ir::Operand::Var(self.jit_mod.last_inst_idx()));
-                }
+                safepoint = Some(inst.safepoint().unwrap());
                 break;
             }
         }
+        // If we don't find a safepoint here something has gone wrong with the AOT IR.
+        let safepoint = safepoint.unwrap();
+        let (rec, _) = AOT_STACKMAPS
+            .as_ref()
+            .unwrap()
+            .get(usize::try_from(safepoint.id).unwrap());
+
+        debug_assert!(safepoint.lives.len() == rec.live_vars.len());
+        for idx in 0..safepoint.lives.len() {
+            let aot_op = &safepoint.lives[idx];
+            let input_tyidx = self.handle_type(aot_op.type_(self.aot_mod))?;
+            let load_ti_inst =
+                jit_ir::LoadTraceInputInst::new(u32::try_from(idx).unwrap(), input_tyidx).into();
+            self.jit_mod.push(load_ti_inst)?;
+
+            // Get the location for this input variable.
+            let var = &rec.live_vars[idx];
+            if var.len() > 1 {
```

Comment:
Is this an `assert` in disguise or a real `todo`?

Reply:
[9932886](https://github.com/ykjit/yk/pull/1422/commits/9932886c1f495981805e07b771bd4a1290defc9a)